buffer/src/main/java/io/netty/buffer/ByteVisitor.java

```diff
@@ -0,0 +1,128 @@
+        private final byte byteToFind;
+
+        public IndexOfVisitor(byte byteToFind) {
+            this.byteToFind = byteToFind;
+        }
+
+        @Override
+        public boolean process(byte value) {
+            return value != byteToFind;
+        }
+    }
+
+    /**
+     * A {@link ByteVisitor} which finds the first appearance which is not of a specific byte.
+     */
+    class IndexNotOfVisitor implements ByteVisitor {
+        private final byte byteToNotFind;
+
+        public IndexNotOfVisitor(byte byteToNotFind) {
+            this.byteToNotFind = byteToNotFind;
+        }
+
+        @Override
+        public boolean process(byte value) {
+            return value == byteToNotFind;
+        }
+    }
+
+    /**
+     * Aborts on a {@code NUL (0x00)}.
+     */
+    ByteVisitor FIND_NUL = new IndexOfVisitor((byte) 0);
+
+    /**
+     * Aborts on a non-{@code NUL (0x00)}.
+     */
+    ByteVisitor FIND_NON_NUL = new IndexNotOfVisitor((byte) 0);
+
+    /**
+     * Aborts on a {@code CR ('\r')}.
+     */
+    ByteVisitor FIND_CR = new IndexOfVisitor((byte) '\r');
+
+    /**
+     * Aborts on a non-{@code CR ('\r')}.
+     */
+    ByteVisitor FIND_NON_CR = new IndexNotOfVisitor((byte) '\r');
+
+    /**
+     * Aborts on a {@code LF ('\n')}.
+     */
+    ByteVisitor FIND_LF = new IndexOfVisitor((byte) '\n');
+
+    /**
+     * Aborts on a non-{@code LF ('\n')}.
+     */
+    ByteVisitor FIND_NON_LF = new IndexNotOfVisitor((byte) '\n');
+
+    /**
+     * Aborts on a {@code CR ('\r')} or a {@code LF ('\n')}.
+     */
+    ByteVisitor FIND_CRLF = new ByteVisitor() {
+        @Override
+        public boolean process(byte value) throws Exception {
+            return value != '\r' && value != '\n';
+        }
+    };
+
+    /**
+     * Aborts on a byte which is neither a {@code CR ('\r')} nor a {@code LF ('\n')}.
+     */
+    ByteVisitor FIND_NON_CRLF = new ByteVisitor() {
+        @Override
+        public boolean process(byte value) throws Exception {
+            return value == '\r' || value == '\n';
+        }
+    };
+
+    /**
+     * Aborts on a linear whitespace (a ({@code ' '} or a {@code '\t'}).
+     */
+    ByteVisitor FIND_LINEAR_WHITESPACE = new ByteVisitor() {
+        @Override
+        public boolean process(byte value) throws Exception {
+            return value != ' ' && value != '\t';
+        }
+    };
+
+    /**
+     * Aborts on a byte which is not a linear whitespace (neither {@code ' '} nor {@code '\t'}).
+     */
+    ByteVisitor FIND_NON_LINEAR_WHITESPACE = new ByteVisitor() {
+        @Override
+        public boolean process(byte value) throws Exception {
+            return value == ' ' || value == '\t';
+        }
+    };
+
+    /**
+     * @return {@code true} if the processor wants to continue the loop and handle the next byte in the buffer.
+     *         {@code false} if the processor wants to stop handling bytes and abort the loop.
+     */
+    boolean process(byte value) throws Exception;
```

Comment:
I think I would prefer `visit` but that would break a bunch of stuff...so I'll rename to `ByteProcessor`.